### PR TITLE
Search Box Autocomplete

### PR DESCRIPTION
**Pull request by ryuyu**
This PR adds in search box "autocomplete" and a new api endpoint to support this. @anangaur 
On Front Page:
image

On Packages Page:
image

Current interaction scheme:
Arrow keys up and Down while focused in search box will move up and down the autocomplete list.
Tabbing/esc will close the list.
Enter with result selected/focused will navigate to package details page.

### CODE REVIEW

**Comment by anangaur** [comment]
Nice!! I like black color better than the blue for the package name. Is it possible to append the download count along with the package ID?

E.g.

MyPackage.Sample (\<downloadIcon>21,345)   Description.

Package ID  - black
Rest - Gray

File: src/NuGetGallery/Scripts/gallery/autocomplete.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        _resultsCache.results = ko.observable();
+        $(document).keydown(function (e) {
+            if (e.keyCode === 27) {
+                removeOldAutocompleteResults();
+                e.stopPropagation();
+            }
+        });
+
+        $("#autocomplete-results-container").keydown(function (e) {
+            if (e.keyCode === 40) {
```
Comment (scottbommarito) [inline src/NuGetGallery/Scripts/gallery/autocomplete.js]: I think every keycode should have either a comment or an appropriately named constant. You put comments in the searchBox key which was good, but I don't see comments on all the keycodes.

```
if (e.KeyCode === _downArrow) {
```

or

```
if (e.KeyCode === 40) { // down arrow
```

File: src/NuGetGallery/Scripts/gallery/autocomplete.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    'use strict';
+    var _autocompleteTimeout = 0;
+    var _autocompleteDelay = 100; //ms
+    var _resultsCache = {};
+    var _maxResults = 9;
+    var _lastIndex = 0;
+
+    function hookAutocomplete(maxResults) {
+        _resultsCache.results = ko.observable();
+        $(document).keydown(function (e) {
```
Comment (scottbommarito) [inline src/NuGetGallery/Scripts/gallery/autocomplete.js]: May want to add some more comments through this code in general, we're C# devs not JS devs so some of this stuff goes over our heads a little more than the usual knockout stuff does.

Especially for the stuff below with the timeouts, e.g.
```
// We will only begin loading the autocomplete results X seconds after the user has typed into the autocomplete box.
```

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 3]
```diff
@@ -686,6 +686,26 @@ protected internal virtual Stream ReadPackageFromRequest()
             };
         }
 
+        [HttpGet]
+        [ActionName("PackageDetails")]
+        public virtual async Task<ActionResult> GetPackageDetails(
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: >GetPackageDetails [](start = 48, length = 17)

change the name to indicate that this searches

File: src/NuGetGallery/Scripts/gallery/autocomplete.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        _autocompleteTimeout = 0;
+    }
+
+    function doSearch(maxResults) {
+        var currInput = $("#search").val();
+        if (currInput.length < 1) {
+            return;
+        }
+
+        var requestUrl = "/api/v2/package-ids?partialId=" + currInput + "&semVerLevel=2.0.0";
```
Comment (skofman1) [inline src/NuGetGallery/Scripts/gallery/autocomplete.js]: please test on DEV to see if this works well on APIM

File: src/NuGetGallery/Scripts/gallery/autocomplete.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        });
+    }
+
+    function setupAllAuxData(idList) {
+        var requestUrl = "/api/v2/package-details?searchString=";
+        for (var i = 0; i < idList.length; i++) {
+            var tempId = idList[i];
+            var searchData = _resultsCache[safeId(tempId)];
+            if (typeof searchData() == "string") {
+                requestUrl += "packageId:" + idList[i] + " ";
```
Comment (joelverhagen) [inline src/NuGetGallery/Scripts/gallery/autocomplete.js]: nit: `packageid`

File: src/NuGetGallery/Scripts/gallery/autocomplete.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+                    }
+                }
+
+                setupAllAuxData(data.slice(0, maxResults));
+            }
+        });
+    }
+
+    function setupAllAuxData(idList) {
+        var requestUrl = "/api/v2/package-details?searchString=";
```
Comment (joelverhagen) [inline src/NuGetGallery/Scripts/gallery/autocomplete.js]: `q`

Comment (joelverhagen) [reply src/NuGetGallery/Scripts/gallery/autocomplete.js]: (instead of `searchString`

**Comment by joelverhagen** [comment]
Have we done an accessibility pass on this?

Could you ask @anangaur if we should talk to UX exports about this change?

**Comment by joelverhagen** [comment]
Accessibility then :shipit:

**Comment by maartenba** [comment]
A new endpoint under /api/v2. Can it somehow be documented? Do third party servers need it?

**Comment by joelverhagen** [comment]
@maartenba, this endpoint is an implementation detail for the auto-complete box and will not be used by the client. 

> Can it somehow be documented?

@ryuyu, @skofman1 - feel free to chime in here, but seeing as even the parts of the V2 protocol that the client uses are not documented, it doesn't make sense here.

> Do third party servers need it?

Nope. Unless you want to implement an auto-complete search box in the same way... but even that it's an implementation detail.

**Comment by anangaur** [comment]
I am not in favor of documenting any V2 APIs. If required, we can have a work item tracking a corresponding V3 API but not something we can work on, in near term.

**Comment by maartenba** [comment]
Thanks! Not a requirement imo but given it suddenly appears on the API/v2 path I just had to ask.

**Comment by xavierdecoster** [comment]
@anangaur @joelverhagen please bare in mind that properly rolling out APIM in front of the gallery may require "some" v2 API docs going forward.

Reason: APIM efforts mandate we define these API endpoints, and their expected behavior in APIM config rules (e.g. request throttling, redirect to HTTPS, etc). 
Missing an endpoint under `/api/v2` in APIM config rules can have unintended side-effects.
On the other hand, the very act of configuring APIM rules in a way is documenting the API v2 endpoints ;)

**Pull request merged**